### PR TITLE
terraform@0.14 0.14.11 (new formula)

### PR DESCRIPTION
**Pull request by chenrui333**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----

previous effort #75511

### CODE REVIEW

File: Formula/terraform@0.14.rb
Comment (carlocab) [inline Formula/terraform@0.14.rb:21]:
```suggestion
    system "go", "build", *std_go_args(ldflags: "-s -w"), "-o", bin/"terraform"
```

**Review by carlocab — CHANGES_REQUESTED** [review]

File: Formula/terraform@0.14.rb
Comment (chenrui333) [reply Formula/terraform@0.14.rb]: keep missing it. 😅

**Review by SMillerDev — CHANGES_REQUESTED** [review]
We still need to have some indication they'll actively maintain it. "One release does not supported make" -Aristotle

**Comment by dsnellgrove** [comment]
While Hashicorp does not appear to have a standing opensource support policy for old versions, they also lack a support policy for the latest releases as well. If an opensource release support policy for a version is required for inclusion in Brew terraform should not be included in Brew at all.

While I agree that empirical evidence of support is poor evidence, they have a strong track record of providing public bugfixes for issues in versions that were released in years ago that match their enterprise support policy. Quick examples include:
* Fixes for 0.11-0.14 released today https://github.com/hashicorp/terraform/releases/tag/v0.11.15
* Fixes for 0.12 and 0.13 released after the availability of 0.14 https://github.com/hashicorp/terraform/releases/tag/v0.12.30

I absolutely understand your stance on only wanting to include supported software Brew. Dead software brings security and bug nightmares! However, the measuring stick by which "is it dead?" being a support policy doesn't seem to work in this case where there is no opensource support policy for _any_ version, even latest, but clear evidence that multiple older versions _are_ being supported. A clarification for how "supported" should be determined would be helpful in circumstances such as this.

**Comment by carlocab** [comment]
I'm personally convinced that `terraform@0.14` is maintained [*] However, one thing that might help to assuage Sean's concerns is to get confirmation from the terraform devs that they intend to continue to release updates to the 0.14 branch for at least the foreseeable future.

Maybe asking in HashiCorp's community forum might be the way to do this?

[*] e.g. there seems to be a [v0.14.12](https://github.com/hashicorp/terraform/blob/v0.14/CHANGELOG.md) planned, and I'm pretty sure Aristotle also said this is enough to establish support.

**Comment by SMillerDev** [comment]
> A clarification for how "supported" should be determined would be helpful in circumstances such as this.

For the current version ongoing development would be enough to see the version isn't dead. But obviously that's not so easy to see for older versions. So usually projects like [php](https://www.php.net/supported-versions.php), [postgres](https://www.postgresql.org/support/versioning/) or [go](https://golang.org/doc/devel/release.html#policy) publish how long they are planning to support something. I understand terraform has this for enterprise users, but since homebrew isn't an enterprise we can't really read that statement and make sure we clean up old versions.

**Comment by SMillerDev** [comment]
To be clear, I'm not debating if 0.14 is maintained. But unless we know that they'll maintain it publicly beyond, e.g. next week I don't think we should be shipping it.

**Comment by carlocab** [comment]
If you agree that it's maintained, then to me that sounds sufficient for the criteria [we currently state](https://docs.brew.sh/Versions#acceptable-versioned-formulae).

If there are more stringent criteria you have in mind, I'm open to that too, but I think we should at least be clear on what they are and make sure the docs reflect them accurately.

**Comment by SMillerDev** [comment]
While we might think it's maintained there is no official word about it (unless you're an enterprise user). So in that case I think we should set the depreciation date to the planned release date for 0.14.2 (and/or make a comment that it should be checked again at that time)

**Comment by carlocab** [comment]
Sounds like we should just ask upstream then.

I think we may also wish to clarify this part of the docs, as it seems the most relevant to this discussion:

>Upstream should have a release branch for each formula version, and release security updates for each version when necessary.

In my view this formula satisfies this requirement, but it seems you want explicit promises from upstream for this to be satisfied, so the docs should probably reflect that.

**Comment by RayBB** [comment]
Having 0.14 is valuable because terraform doesn't let you upgrade directly from 0.13 to 0.15 and requires using 0.14 be used to upgrade from 0.13

Just adding some color on why it's valuable and likely to be useful for folks

https://github.com/Homebrew/discussions/discussions/1371

**Comment by SMillerDev** [comment]
If you think 0.14 is valuable, could you get someone from hashicorp to state what their open source support policy is?

**Comment by carlocab** [comment]
> If you think 0.14 is valuable, could you get someone from hashicorp to state what their open source support policy is?

In particular, how long they intend to provide version updates to the 0.14.x branch would be useful.

**Comment by RayBB** [comment]
Well I opened an issue that you can see hashicorp/terraform#28538 in hopes that they state something official.

As was mentioned above. Version 0.11 that was first released in 2017 just had an update this week. Just like 0.12, 0.13, 0.14, 0.15. 
I understand why you don't want old software that is no longer supported added here. Version 0.14 is receiving updates on cadence with both older and newer major versions. We can see that they are supporting it at least as of two days ago. Does having the company just say "we'll support it for 6 months" really change anything? They could just not follow up on their statement.

Is their track record of supporting old versions for several years and having publicly stated they plan a future release not enough? “Actions speak louder than words” - Aristotle

**Comment by SMillerDev** [comment]
> Does having the company just say "we'll support it for 6 months" really change anything?

Yes, because it makes it easier for homebrew to see if effort should be taken to keep versions of software running (if it's deprecated next week I won't be writing a patch for support with the latest Go version). And it allows us to warn users in advance instead of just instantly removing the file when it becomes unsupported.

Homebrew has a limited number of maintainers to maintain an ever growing software collection. Rules like this are needed to make this feasible.

**Comment by apparentlymart** [comment]
Hi all! I work on the Terraform Core team which maintains the upstream that you're talking about packaging here.

We typically have two development tracks at a time:
* Maintenance of the current version, which is usually (but not always) just backports of bug fixes. Currently that's v0.15.
* Development of the next version, where most features will land.

For a short time after a new major release (which is what we're currently calling releases like v0.15) we may also backport high-impact bug fixes to the previous release, particularly if they'll help folks wit upgrading to the latest. v0.14 is currently in that situation because v0.15 is only a couple weeks old, but of course whether there will be any further releases will depend on whether anything urgent comes up. Based on previous releases, I'd expect there to be only a couple more v0.14 releases at most, because our primary maintenance efforts have now shifted to v0.15.

This week there was [a particularly important event related to security bulletin HCSEC-2021-12](https://discuss.hashicorp.com/t/terraform-updates-for-hcsec-2021-12/23570/2), which has prompted us to issue new releases going back all the way to the 0.11 line as a special exception. That isn't a typical practice, but we will make exceptions to the usual policy for security-related changes. I wouldn't recommend deciding general policy based on this exceptional situation, because this is the first time any such event has occurred through Terraform's history.

I don't have any specific advice about what would be valuable or practical for Homebrew to support, though. All historical versions will always be available from our official release channels anyway, so it could be reasonable to have a policy that the Homebrew releases are targeted at folks who stay reasonably up-to-date while folks who want older versions should get them other ways, if that policy makes the best use of Homebrew maintainers' time. I certainly wouldn't demand anyone working in their spare time spend time on anything they don't find valuable.

Terraform's release engineering team does maintain [a HashiCorp-specific homebrew tap](https://www.hashicorp.com/blog/announcing-hashicorp-homebrew-tap), which re-publishes the official releases in a Homebrew-compatible way, and so that could potentially be a good compromise for folks who need older versions of Terraform that the Homebrew community doesn't have the resources to support.

I hope this information helps with deciding a suitable policy!

**Comment by SMillerDev** [comment]
> Based on previous releases, I'd expect there to be only a couple more v0.14 releases at most, because our primary maintenance efforts have now shifted to v0.15.

> Terraform's release engineering team does maintain a HashiCorp-specific homebrew tap, which re-publishes the official releases in a Homebrew-compatible way, and so that could potentially be a good compromise for folks who need older versions of Terraform that the Homebrew community doesn't have the resources to support.

Based on this I'd say we only keep the latest version in brew and defer to the tap for any older versions. @Homebrew/core any objections?

**Comment by MikeMcQuaid** [comment]
> Based on this I'd say we only keep the latest version in brew and defer to the tap for any older versions. @Homebrew/core any objections?

Works for me 👍🏻

**Comment by dsnellgrove** [comment]
As some random user that has benefitted from versioned formula for terraform in the past:

* While the vendor tap doesn't have formula for older versions either, I think it's reasonable for people to ask Hashcorp to add that feature if they want it (Hashicorp has declined in the past https://github.com/hashicorp/homebrew-tap/issues/79)
* Hashicorp also provides single-binary downloads of older versions that users can grab in a pinch to finish upgrading to latest (https://releases.hashicorp.com/terraform/)
* Users can build their own tap with the prior versions (https://docs.brew.sh/Manpage#extract---version---force-formula-tap) and it even puts a version number on the binary name by default so many can be installed at once (nice!)

I would find it totally reasonable for Homebrew to remove all the older versions and just ship the latest. It's a wonderful convenience to the community as it is. Ya'll are awesome and so is the work you do, keep doing the most important things :)

**Comment by apparentlymart** [comment]
Oh yes... sorry I was speaking beyond my knowledge when I asserted that older versions of Terraform would be available from the HashiCorp tap. I'm not directly involved in the tap-population mechanism and so I made an incorrect assumption. The discussions in the hashicorp/homebrew-tap repository should have more weight than my comments about it.

Our canonical source of release packages is indeed https://releases.hashicorp.com/terraform/ , which contains zip files that each contain a single executable which should be immediately runnable once extracted. The other distribution channels are derived from those.

As one of my colleagues noted in hashicorp/homebrew-tap#79, another option is to install [`tfenv`](https://github.com/tfutils/tfenv) using homebrew and then use `tfenv` to manage Terraform versions. Our team doesn't maintain `tfenv` but its source of available versions is https://releases.hashicorp.com/terraform/ and so it can install and arrange to run run all of the releases published there.

**Comment by albgus** [comment]
@apparentlymart For me this seems to be a quite major issue, since each v0.x releases are not compatible with each-other. If there is no older versions provided, how is one expected to perform upgrades of terraform when working with multiple workspaces?

To me this seems like a major issue with the terraform formula and if neither terraforms official tap or the community build will offer this functionality I guess the only actual solution is to stop using homebrew for Terraform.

**Comment by dsnellgrove** [comment]
@albgus As listed here https://github.com/Homebrew/homebrew-core/pull/75981#issuecomment-829218317 you're probably looking for the [extract --version=](https://docs.brew.sh/Manpage#extract---version---force-formula-tap) feature of Terraform to use multiple/older versions of a formula, or you can use tfenv to install multiple/older versions of just Terraform (and tfenv can be installed by homebrew)

**Pull request closed without merging**
